### PR TITLE
Implementation-of-flip-69

### PR DESCRIPTION
**Pull request by satyamakgec**
It addresses https://github.com/onflow/flips/blob/fungibleToken-standard-improvement/flips/20230206-fungible-token-vault-type-discovery.md

TODO:

- [ ] Update package.json
- [x] fix js-test framework issue

### CODE REVIEW

File: contracts/FungibleToken.cdc [hunk 1]
```diff
@@ -109,6 +109,27 @@ pub contract interface FungibleToken {
...
+        /// Below is referenced from the FLIP #69 https://github.com/onflow/flips/blob/main/flips/20230206-fungible-token-vault-type-discovery.md
+        /// 
+        /// Returns the type of implementing resource i.e If `FlowToken.Vault` implements
+        /// this then it would return `[Type<@FlowToken.Vault>()]` and if any custom receiver
+        /// uses the default implementation then it would return empty array as its parent
+        /// resource doesn't conform with the `FungibleToken.Vault` resource.
+        ///
+        /// @return list of supported vault types by the implemented resource.
+        /// 
+        pub fun getSupportedVaultTypes() :[Type] {
```
Comment (joshuahannan) [inline contracts/FungibleToken.cdc:122]:
```suggestion
        pub fun getSupportedVaultTypes(): [Type] {
```

File: contracts/FungibleToken.cdc [hunk 2]
```diff
@@ -109,6 +109,27 @@ pub contract interface FungibleToken {
         /// @param from: The Vault resource containing the funds that will be deposited
         ///
         pub fun deposit(from: @Vault)
+
+        /// Below is referenced from the FLIP #69 https://github.com/onflow/flips/blob/main/flips/20230206-fungible-token-vault-type-discovery.md
+        /// 
+        /// Returns the type of implementing resource i.e If `FlowToken.Vault` implements
```
Comment (joshuahannan) [inline contracts/FungibleToken.cdc:115]:
```suggestion
        /// Returns the Vault types that the the receiver is able to accept in its `deposit` method
        /// By default, it returns the concrete type of the implementing resource if it is a subtype of `FungibleToken.Vault`
        /// i.e If `FlowToken.Vault` implements
```

**Review by joshuahannan — COMMENTED** [review]
I looked at the cadence code. Looks good! Can you also add implementations for the switchboard and token forwarders, and can you add a transfer transaction that accepts an address and public path as a parameter, gets the receiver reference from that path, then only deposits the tokens if the `receiver.getSupportedVaultTypes()` supports the type of the vault that the transaction is attempting to transfer? If it doesn't just deposit the vault back into the owner's vault

**Review by franklywatson — APPROVED** [review]

**Review by joshuahannan — APPROVED** [review]
I'm good with it. Lets share this on discord to get some community input. Make sure to say that we want to upgrade this on testnet and mainnet soon

File: contracts/FungibleToken.cdc [hunk 3]
```diff
@@ -109,6 +109,30 @@ pub contract interface FungibleToken {
...
+        /// this then it would return `[Type<@FlowToken.Vault>()]` and if any custom receiver
+        /// uses the default implementation then it would return empty array as its parent
+        /// resource doesn't conform with the `FungibleToken.Vault` resource.
+        ///
+        /// Custom receiver implementations are expected to upgrade their contracts to add an implementation
+        /// that supports this method because it is very valuable for various applications to have.
+        ///
+        /// @return list of supported deposit vault types by the implementing resource.
+        /// 
+        pub fun getSupportedVaultTypes(): [Type] {
```
Comment (joshuahannan) [inline contracts/FungibleToken.cdc:125]: Thinking about this more, I would like the return type of this to be `{Type: Bool}`. You can still iterate through it with `return.keys`, and it makes checking a specific type much easier. Do you think you'll be able to make that change in time for the upgrade next week?

**Comment by joshuahannan** [comment]
@satyamakgec You'll need to pull the latest from master before you push your new changes since it is out of date

**Comment by satyamakgec** [comment]
@joshuahannan latest pull have taken and the return type also changed

**Comment by satyamakgec** [comment]
May need to fix the Go tests to make the CI happy

**Review by joshuahannan — COMMENTED** [review]
I'll fix the Go tests right now

**Review by joshuahannan — APPROVED** [review]

**Pull request merged**